File: mailblastr-go/emails_test.go
package mailblastr

import (
	"context"
	"encoding/json"
	"net/http"
	"strings"
	"testing"
)

func TestEmailsSend(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			t.Errorf("method = %s, want POST", r.Method)
		}
		if r.URL.Path != "/emails" {
			t.Errorf("path = %s, want /emails", r.URL.Path)
		}
		assertAuth(t, r)
		if ct := r.Header.Get("Content-Type"); ct != "application/json" {
			t.Errorf("Content-Type = %q, want application/json", ct)
		}
		body := decodeBody(t, r)
		if body["from"] != "Acme <hello@yourdomain.com>" {
			t.Errorf("from = %v", body["from"])
		}
		to, ok := body["to"].([]any)
		if !ok || len(to) != 1 || to[0] != "user@example.com" {
			t.Errorf("to = %v", body["to"])
		}
		if body["subject"] != "Hello" {
			t.Errorf("subject = %v", body["subject"])
		}
		if body["html"] != "<p>Hi</p>" {
			t.Errorf("html = %v", body["html"])
		}
		if _, present := body["text"]; present {
			t.Error("empty text should be omitted from the body")
		}
		w.Write([]byte(`{"id":"em_123"}`))
	})

	sent, err := client.Emails.Send(&SendEmailRequest{
		From:    "Acme <hello@yourdomain.com>",
		To:      []string{"user@example.com"},
		Subject: "Hello",
		Html:    "<p>Hi</p>",
	})
	if err != nil {
		t.Fatalf("Send: %v", err)
	}
	if sent.Id != "em_123" {
		t.Errorf("Id = %q, want em_123", sent.Id)
	}
}

func TestEmailsSendWithOptionsIdempotencyKey(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("Idempotency-Key"); got != "order-123" {
			t.Errorf("Idempotency-Key = %q, want order-123", got)
		}
		w.Write([]byte(`{"id":"em_1"}`))
	})

	_, err := client.Emails.SendWithOptions(context.Background(),
		&SendEmailRequest{From: "a@b.com", To: []string{"c@d.com"}, Subject: "s", Text: "t"},
		&RequestOptions{IdempotencyKey: "order-123"})
	if err != nil {
		t.Fatalf("SendWithOptions: %v", err)
	}
}

// The documented bound is 1-255 characters, measured AFTER the server trims
// the value (api_idempotency.key is VARCHAR(255)) — 255, not 256. The constant
// is exported so the rule is discoverable; the key itself is sent verbatim and
// the SERVER answers an out-of-range one with 400 invalid_idempotency_key, so
// this package never pre-checks it.
func TestIdempotencyKeyIsSentVerbatimAndBoundedByTheServer(t *testing.T) {
	if IdempotencyKeyMaxLen != 255 {
		t.Fatalf("IdempotencyKeyMaxLen = %d, want 255", IdempotencyKeyMaxLen)
	}

	tooLong := strings.Repeat("k", IdempotencyKeyMaxLen+1)
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("Idempotency-Key"); got != tooLong {
			t.Errorf("Idempotency-Key length = %d, want %d (sent verbatim)", len(got), len(tooLong))
		}
		w.Write([]byte(`{"id":"em_1"}`))
	})

	_, err := client.Emails.SendWithOptions(context.Background(),
		&SendEmailRequest{From: "a@b.com", To: []string{"c@d.com"}, Subject: "s", Text: "t"},
		&RequestOptions{IdempotencyKey: tooLong})
	if err != nil {
		t.Fatalf("the SDK must not reject the key locally: %v", err)
	}
}

func TestEmailsListPagination(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			t.Errorf("method = %s, want GET", r.Method)
		}
		if r.URL.Path != "/emails" {
			t.Errorf("path = %s, want /emails", r.URL.Path)
		}
		q := r.URL.Query()
		if q.Get("limit") != "20" || q.Get("after") != "em_9" {
			t.Errorf("query = %v", q)
		}
		w.Write([]byte(`{"object":"list","has_more":true,"data":[{"object":"email","id":"em_10","from":"a@b.com","to":["c@d.com"],"subject":"s","last_event":"delivered","created_at":"2026-07-06T00:00:00Z"}]}`))
	})

	list, err := client.Emails.List(&ListParams{Limit: 20, After: "em_9"})
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if !list.HasMore || len(list.Data) != 1 || list.Data[0].Id != "em_10" {
		t.Errorf("unexpected list: %+v", list)
	}
}

func TestEmailsListFilteredStatusAndSearch(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		if q.Get("status") != "bounced" {
			t.Errorf("status = %q, want bounced", q.Get("status"))
		}
		if q.Get("search") != "invoice" {
			t.Errorf("search = %q, want invoice", q.Get("search"))
		}
		if q.Get("domain_id") != "dom_1" || q.Get("campaign_id") != "cmp_1" {
			t.Errorf("query = %v", q)
		}
		w.Write([]byte(`{"object":"list","has_more":false,"data":[{"object":"email","id":"em_1","domain_id":"dom_1","campaign_id":"cmp_1","automation_id":null,"last_event":"bounced","to":["a@b.com"],"from":"c@d.com","subject":"invoice","created_at":"2026-08-08T00:00:00Z"}]}`))
	})

	list, err := client.Emails.ListFiltered(&ListEmailsRequest{
		Status:     "bounced",
		Search:     "invoice",
		DomainId:   "dom_1",
		CampaignId: "cmp_1",
	})
	if err != nil {
		t.Fatalf("ListFiltered: %v", err)
	}
	got := list.Data[0]
	if got.DomainId != "dom_1" || got.CampaignId != "cmp_1" {
		t.Errorf("origin fields not decoded: %+v", got)
	}
}

func TestEmailsSources(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet || r.URL.Path != "/emails/sources" {
			t.Errorf("%s %s, want GET /emails/sources", r.Method, r.URL.Path)
		}
		w.Write([]byte(`{"object":"list","has_more":false,"data":[{"kind":"campaign","id":"cmp_1","name":"Launch","subject":"Hi","status":"sent","total":10,"sent":10,"delivered":9,"opened":4,"clicked":1,"replied":0,"failed":0,"last_sent_at":"2026-08-08T10:00:00.000Z"}]}`))
	})

	list, err := client.Emails.Sources()
	if err != nil {
		t.Fatalf("Sources: %v", err)
	}
	src := list.Data[0]
	if src.Kind != "campaign" || src.Delivered != 9 || src.LastSentAt == "" {
		t.Errorf("unexpected source row: %+v", src)
	}
}

func TestReceivingListAddresses(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/emails/receiving/addresses" {
			t.Errorf("path = %s, want /emails/receiving/addresses", r.URL.Path)
		}
		w.Write([]byte(`{"object":"list","has_more":false,"data":[{"address":"hi@x.com","total":12,"replies":3,"interested":1,"last_received_at":"2026-08-08T10:00:00.000Z"}]}`))
	})

	list, err := client.Emails.Receiving.ListAddresses()
	if err != nil {
		t.Fatalf("ListAddresses: %v", err)
	}
	if list.Data[0].Address != "hi@x.com" || list.Data[0].Total != 12 {
		t.Errorf("unexpected row: %+v", list.Data[0])
	}
}

func TestBatchSend(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/emails/batch" {
			t.Errorf("%s %s, want POST /emails/batch", r.Method, r.URL.Path)
		}
		w.Write([]byte(`{"data":[{"id":"em_1"},{"id":"em_2"}]}`))
	})

	res, err := client.Batch.Send([]*SendEmailRequest{
		{From: "a@b.com", To: []string{"x@y.com"}, Subject: "1", Text: "one"},
		{From: "a@b.com", To: []string{"z@y.com"}, Subject: "2", Text: "two"},
	})
	if err != nil {
		t.Fatalf("Batch.Send: %v", err)
	}
	if len(res.Data) != 2 || res.Data[1].Id != "em_2" {
		t.Errorf("unexpected batch response: %+v", res)
	}
}

func TestReceivingGetAttachmentRawBytes(t *testing.T) {
	raw := []byte{0x25, 0x50, 0x44, 0x46} // %PDF
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/emails/receiving/rcv_1/attachments/att_1" {
			t.Errorf("path = %s", r.URL.Path)
		}
		assertAuth(t, r)
		w.Header().Set("Content-Type", "application/pdf")
		w.Write(raw)
	})

	got, err := client.Emails.Receiving.GetAttachment("rcv_1", "att_1")
	if err != nil {
		t.Fatalf("GetAttachment: %v", err)
	}
	if string(got) != string(raw) {
		t.Errorf("bytes = %v, want %v", got, raw)
	}
}

// A send that references a template must OMIT an empty `from`/`subject` so the
// template's own values fill in. The API keys off key PRESENCE
// (lib/send/resolve_template.ts: `body.from != null ? body.from : published.from`),
// so `"from":""` counted as "supplied": the send was rejected with 422
// missing_required_field, and a supplied-from/blank-subject payload went out
// with an EMPTY subject instead of the template's.
func TestTemplateSendOmitsEmptyFromAndSubject(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		body := decodeBody(t, r)
		if _, present := body["from"]; present {
			t.Errorf("empty from must be omitted for a template send, got %#v", body["from"])
		}
		if _, present := body["subject"]; present {
			t.Errorf("empty subject must be omitted for a template send, got %#v", body["subject"])
		}
		tpl, ok := body["template"].(map[string]any)
		if !ok || tpl["id"] != "tpl_1" {
			t.Errorf("template = %#v", body["template"])
		}
		w.Write([]byte(`{"id":"em_1"}`))
	})

	_, err := client.Emails.Send(&SendEmailRequest{
		To:       []string{"delivered@mailblastr.dev"},
		Template: &TemplateRef{Id: "tpl_1", Variables: map[string]any{"name": "Ada"}},
	})
	if err != nil {
		t.Fatalf("Send: %v", err)
	}
}

// The flat `template_id` form gets the same treatment.
func TestTemplateIdSendOmitsEmptyFromAndSubject(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		body := decodeBody(t, r)
		if _, present := body["from"]; present {
			t.Error("empty from must be omitted when template_id is set")
		}
		if _, present := body["subject"]; present {
			t.Error("empty subject must be omitted when template_id is set")
		}
		if body["template_id"] != "tpl_1" {
			t.Errorf("template_id = %#v", body["template_id"])
		}
		w.Write([]byte(`{"id":"em_1"}`))
	})

	if _, err := client.Emails.Send(&SendEmailRequest{
		To:         []string{"delivered@mailblastr.dev"},
		TemplateId: "tpl_1",
	}); err != nil {
		t.Fatalf("Send: %v", err)
	}
}

// An explicit From/Subject still overrides the template's, so both keys are
// present when the caller supplies them.
func TestTemplateSendKeepsExplicitFromAndSubject(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		body := decodeBody(t, r)
		if body["from"] != "Acme <hello@yourdomain.com>" || body["subject"] != "Override" {
			t.Errorf("from = %#v, subject = %#v", body["from"], body["subject"])
		}
		w.Write([]byte(`{"id":"em_1"}`))
	})

	if _, err := client.Emails.Send(&SendEmailRequest{
		From:       "Acme <hello@yourdomain.com>",
		Subject:    "Override",
		To:         []string{"delivered@mailblastr.dev"},
		TemplateId: "tpl_1",
	}); err != nil {
		t.Fatalf("Send: %v", err)
	}
}

// Without a template the two keys are ALWAYS sent, empty included: the API
// documents `subject: ""` as a legal empty subject (only null/absent is
// "missing"), and a missing `from` must still surface as the server's
// missing_required_field rather than being silently reshaped by the SDK.
func TestNonTemplateSendAlwaysSendsFromAndSubject(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		body := decodeBody(t, r)
		from, hasFrom := body["from"]
		subject, hasSubject := body["subject"]
		if !hasFrom || from != "" {
			t.Errorf("from = %#v (present=%v), want present and empty", from, hasFrom)
		}
		if !hasSubject || subject != "" {
			t.Errorf("subject = %#v (present=%v), want present and empty", subject, hasSubject)
		}
		w.Write([]byte(`{"id":"em_1"}`))
	})

	if _, err := client.Emails.Send(&SendEmailRequest{
		To:   []string{"delivered@mailblastr.dev"},
		Text: "no template here",
	}); err != nil {
		t.Fatalf("Send: %v", err)
	}
}

// A nested Template carrying neither Id nor Alias references NOTHING — the
// server's extractor lets a non-null `template` shadow `template_id` — so
// from/subject must keep being sent.
func TestEmptyTemplateRefShadowsTemplateId(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		body := decodeBody(t, r)
		if _, present := body["from"]; !present {
			t.Error("from must still be sent: an id-less template references nothing")
		}
		if _, present := body["subject"]; !present {
			t.Error("subject must still be sent: an id-less template references nothing")
		}
		w.Write([]byte(`{"id":"em_1"}`))
	})

	if _, err := client.Emails.Send(&SendEmailRequest{
		To:         []string{"delivered@mailblastr.dev"},
		TemplateId: "tpl_1",
		Template:   &TemplateRef{Variables: map[string]any{"name": "Ada"}},
	}); err != nil {
		t.Fatalf("Send: %v", err)
	}
}

// Batch items follow the same rule.
func TestBatchTemplateItemOmitsEmptyFromAndSubject(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		var items []map[string]any
		if err := json.NewDecoder(r.Body).Decode(&items); err != nil {
			t.Fatalf("decode batch body: %v", err)
		}
		if len(items) != 1 {
			t.Fatalf("items = %d, want 1", len(items))
		}
		if _, present := items[0]["from"]; present {
			t.Error("empty from must be omitted for a template batch item")
		}
		if _, present := items[0]["subject"]; present {
			t.Error("empty subject must be omitted for a template batch item")
		}
		w.Write([]byte(`{"data":[{"id":"em_1"}]}`))
	})

	if _, err := client.Batch.SendEmails([]*BatchEmailRequest{
		{To: []string{"delivered@mailblastr.dev"}, Template: &TemplateRef{Alias: "welcome"}},
	}); err != nil {
		t.Fatalf("SendEmails: %v", err)
	}
}

// A batch larger than BatchSyncMax is QUEUED, not sent: HTTP 202 with
// `queued`/`queued_count` alongside the reserved ids. Callers must be able to
// tell that apart from an inline send.
func TestBatchQueuedResponse(t *testing.T) {
	if BatchSyncMax != 40 {
		t.Fatalf("BatchSyncMax = %d, want 40", BatchSyncMax)
	}
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusAccepted)
		w.Write([]byte(`{"data":[{"id":"em_1"},{"id":"em_2"}],"queued":true,"queued_count":2}`))
	})

	res, err := client.Batch.SendEmails([]*BatchEmailRequest{
		{From: "a@b.com", To: []string{"delivered@mailblastr.dev"}, Subject: "1", Text: "one"},
		{From: "a@b.com", To: []string{"delivered@mailblastr.dev"}, Subject: "2", Text: "two"},
	})
	if err != nil {
		t.Fatalf("SendEmails: %v", err)
	}
	if !res.Queued || res.QueuedCount != 2 || len(res.Data) != 2 {
		t.Errorf("queued batch not surfaced: %+v", res)
	}
}

// An inline (200) batch must NOT look queued.
func TestBatchInlineResponseIsNotQueued(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`{"data":[{"id":"em_1"}]}`))
	})

	res, err := client.Batch.SendEmails([]*BatchEmailRequest{
		{From: "a@b.com", To: []string{"delivered@mailblastr.dev"}, Subject: "1", Text: "one"},
	})
	if err != nil {
		t.Fatalf("SendEmails: %v", err)
	}
	if res.Queued || res.QueuedCount != 0 {
		t.Errorf("inline batch reported as queued: %+v", res)
	}
}

File: mailblastr-php/tests/cases/EmailsTest.php
<?php

declare(strict_types=1);

// ---- emails->send ----
[$mb, $t] = make_client();
$t->queue(200, ['id' => 'em_1']);
$res = $mb->emails->send([
    'from' => 'Acme <hello@yourdomain.com>',
    'to' => ['user@example.com'],
    'subject' => 'Hello',
    'html' => '<p>Hi</p>',
], ['idempotencyKey' => 'order-123']);
check_same('emails.send: method', 'POST', $t->last()['method']);
check_same('emails.send: path', '/emails', $t->lastPath());
check_same('emails.send: body', [
    'from' => 'Acme <hello@yourdomain.com>',
    'to' => ['user@example.com'],
    'subject' => 'Hello',
    'html' => '<p>Hi</p>',
], $t->lastJson());
check_auth('emails.send', $t->last());
check('emails.send: idempotency header', in_array('Idempotency-Key: order-123', $t->last()['headers'], true));
check_same('emails.send: response', ['id' => 'em_1'], $res);

// The documented bound is 1-255 characters measured AFTER the server trims the
// value (VARCHAR(255)) — 255, not 256. The constant is exported so the rule is
// discoverable; the key itself is sent verbatim and the server is the authority
// (400 invalid_idempotency_key), so an over-long key is NOT rejected here.
check_same('client: idempotency key max length', 255, \Mailblastr\Client::IDEMPOTENCY_KEY_MAX_LENGTH);
$t->queue(200, ['id' => 'em_2']);
$over = str_repeat('k', \Mailblastr\Client::IDEMPOTENCY_KEY_MAX_LENGTH + 1);
$mb->emails->send(['from' => 'a@x.com', 'to' => 'u@e.com', 'subject' => 's', 'text' => 'b'], ['idempotencyKey' => $over]);
check('emails.send: over-long idempotency key is left to the server', in_array('Idempotency-Key: ' . $over, $t->last()['headers'], true));

// A key the API accepts must never be dropped on the way out. '0' and 0 are
// valid 1-character keys, but PHP's empty() is true for both — gating the header
// on !empty() silently turned an idempotent send into a plain one, so an
// automatic 429/503 retry could deliver the same email twice.
$idemHeaders = static fn (array $request): array => array_values(array_filter(
    $request['headers'],
    static fn (string $h): bool => str_starts_with($h, 'Idempotency-Key:')
));
foreach ([['0', "'0'"], [0, '0 (int)'], [12345, '12345 (int)']] as [$key, $label]) {
    $t->queue(200, ['id' => 'em_3']);
    $mb->emails->send(
        ['from' => 'a@x.com', 'to' => 'delivered@mailblastr.dev', 'subject' => 's', 'text' => 'b'],
        ['idempotencyKey' => $key],
    );
    check_same(
        "emails.send: idempotency key {$label} is sent verbatim",
        ['Idempotency-Key: ' . $key],
        $idemHeaders($t->last()),
    );
}
// An absent, empty or non-scalar key sends no header at all (the API rejects an
// empty one with 400 invalid_idempotency_key).
foreach ([[null, 'null'], ['', "''"], [[], 'array']] as [$key, $label]) {
    $t->queue(200, ['id' => 'em_4']);
    $mb->emails->send(
        ['from' => 'a@x.com', 'to' => 'delivered@mailblastr.dev', 'subject' => 's', 'text' => 'b'],
        ['idempotencyKey' => $key],
    );
    check_same("emails.send: idempotency key {$label} sends no header", [], $idemHeaders($t->last()));
}

// ---- batch->send (JSON list body) ----
[$mb, $t] = make_client();
$t->queue(200, ['data' => [['id' => 'em_1'], ['id' => 'em_2']]]);
$mb->batch->send([
    ['from' => 'a@x.com', 'to' => 'u1@e.com', 'subject' => 's1', 'text' => 'b1'],
    ['from' => 'a@x.com', 'to' => 'u2@e.com', 'subject' => 's2', 'text' => 'b2'],
]);
check_same('batch.send: method', 'POST', $t->last()['method']);
check_same('batch.send: path', '/emails/batch', $t->lastPath());
check('batch.send: body is a JSON array', str_starts_with((string) $t->last()['body'], '['));

// emails->batch alias hits the same endpoint
$mb->emails->batch([['from' => 'a@x.com', 'to' => 'u@e.com', 'subject' => 's', 'text' => 'b']]);
check_same('emails.batch alias: path', '/emails/batch', $t->lastPath());

// ---- list with pagination ----
[$mb, $t] = make_client();
$mb->emails->list(['limit' => 2, 'after' => 'em_1']);
check_same('emails.list: method', 'GET', $t->last()['method']);
check_same('emails.list: path', '/emails?limit=2&after=em_1', $t->lastPath());
check_same('emails.list: no body on GET', null, $t->last()['body']);

// status/search filters are server-side
$mb->emails->list(['status' => 'bounced', 'search' => 'ada@', 'domain_id' => 'dom_1']);
check_same('emails.list: status/search filters', '/emails?status=bounced&search=ada%40&domain_id=dom_1', $t->lastPath());

// folder is forwarded too (outbox|sent|scheduled|failed; the server 422s
// any other value)
$mb->emails->list(['folder' => 'scheduled']);
check_same('emails.list: folder filter', '/emails?folder=scheduled', $t->lastPath());

// ---- sources (per-source send metrics, not paginated) ----
$mb->emails->sources();
check_same('emails.sources: method', 'GET', $t->last()['method']);
check_same('emails.sources: path', '/emails/sources', $t->lastPath());

// ---- get with a path-traversal id is escaped ----
$mb->emails->get('../api-keys');
check_same('emails.get: traversal id is percent-encoded', '/emails/..%2Fapi-keys', $t->lastPath());

// ---- update (reschedule) + cancel ----
$mb->emails->update('em_1', ['scheduled_at' => '2026-08-01T00:00:00Z']);
check_same('emails.update: method', 'PATCH', $t->last()['method']);
check_same('emails.update: path', '/emails/em_1', $t->lastPath());
check_same('emails.update: body', ['scheduled_at' => '2026-08-01T00:00:00Z'], $t->lastJson());

$mb->emails->cancel('em_1');
check_same('emails.cancel: method', 'POST', $t->last()['method']);
check_same('emails.cancel: path', '/emails/em_1/cancel', $t->lastPath());
check_same('emails.cancel: no body', null, $t->last()['body']);

// ---- sent-email attachments ----
$mb->emails->listAttachments(id: 'em_1');
check_same('emails.listAttachments: path', '/emails/em_1/attachments', $t->lastPath());
$mb->emails->getAttachment(id: 'em_1', attachmentId: 'att_9');
check_same('emails.getAttachment: path', '/emails/em_1/attachments/att_9', $t->lastPath());

// ---- receiving sub-resource ----
[$mb, $t] = make_client();
$mb->emails->receiving->list(['limit' => 5]);
check_same('receiving.list: path', '/emails/receiving?limit=5', $t->lastPath());

$mb->emails->receiving->get('rcv_1');
check_same('receiving.get: path', '/emails/receiving/rcv_1', $t->lastPath());

$mb->emails->receiving->listAddresses();
check_same('receiving.listAddresses: path', '/emails/receiving/addresses', $t->lastPath());

$mb->emails->receiving->listAttachments('rcv_1');
check_same('receiving.listAttachments: path', '/emails/receiving/rcv_1/attachments', $t->lastPath());

$mb->emails->receiving->listAttachments('rcv_1', ['limit' => 10]);
check_same('receiving.listAttachments: pagination', '/emails/receiving/rcv_1/attachments?limit=10', $t->lastPath());

$t->queue(200, 'RAWBYTES');
$bytes = $mb->emails->receiving->getAttachment('rcv_1', 'att_1');
check_same('receiving.getAttachment: raw bytes returned', 'RAWBYTES', $bytes);
check_same('receiving.getAttachment: path', '/emails/receiving/rcv_1/attachments/att_1', $t->lastPath());
check_auth('receiving.getAttachment', $t->last());

$t->queue(200, "From: a@b.c\r\n\r\nbody");
$raw = $mb->emails->receiving->getRaw('rcv_1');
check_same('receiving.getRaw: raw MIME returned', "From: a@b.c\r\n\r\nbody", $raw);
check_same('receiving.getRaw: path', '/emails/receiving/rcv_1/raw', $t->lastPath());

$mb->emails->receiving->forward('rcv_1', ['from' => 'me@yourdomain.com', 'to' => 'team@you.com']);
check_same('receiving.forward: method', 'POST', $t->last()['method']);
check_same('receiving.forward: path', '/emails/receiving/rcv_1/forward', $t->lastPath());
check_same('receiving.forward: body', ['from' => 'me@yourdomain.com', 'to' => 'team@you.com'], $t->lastJson());

$mb->emails->receiving->reply('rcv_1', ['from' => 'me@yourdomain.com', 'text' => 'thanks!']);
check_same('receiving.reply: path', '/emails/receiving/rcv_1/reply', $t->lastPath());

$mb->emails->receiving->remove('rcv_1');
check_same('receiving.remove: method', 'DELETE', $t->last()['method']);
check_same('receiving.remove: path', '/emails/receiving/rcv_1', $t->lastPath());
